pkg/apis/seedmanagement/v1alpha1/defaults.go

```diff
@@ -0,0 +1,149 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	"encoding/json"
+	"fmt"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	"github.com/gardener/gardener/pkg/apis/seedmanagement/helper"
+	configv1alpha1 "github.com/gardener/gardener/pkg/gardenlet/apis/config/v1alpha1"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/utils/pointer"
+)
+
+func addDefaultingFuncs(scheme *runtime.Scheme) error {
+	return RegisterDefaults(scheme)
+}
+
+// SetDefaults_ManagedSeed sets default values for ManagedSeed objects.
+func SetDefaults_ManagedSeed(obj *ManagedSeed) {
+	switch {
+	case obj.Spec.SeedTemplate != nil:
+		setDefaultsSeedSpec(&obj.Spec.SeedTemplate.Spec, obj.Name, obj.Namespace, true)
+	case obj.Spec.Gardenlet != nil:
+		setDefaultsGardenlet(obj.Spec.Gardenlet, obj.Name, obj.Namespace)
+	}
+}
+
+// SetDefaults_GardenletDeployment sets default values for GardenletDeployment objects.
+func SetDefaults_GardenletDeployment(obj *GardenletDeployment) {
+	// Set default replica count
+	if obj.ReplicaCount == nil {
+		obj.ReplicaCount = pointer.Int32Ptr(1)
+	}
+
+	// Set default revision history limit
+	if obj.RevisionHistoryLimit == nil {
+		obj.RevisionHistoryLimit = pointer.Int32Ptr(1)
+	}
+
+	// Set default image
+	if obj.Image == nil {
+		obj.Image = &Image{}
+	}
+
+	// Set default VPA
+	if obj.VPA == nil {
+		obj.VPA = pointer.BoolPtr(true)
+	}
+}
+
+// SetDefaults_Image sets default values for Image objects.
+func SetDefaults_Image(obj *Image) {
+	// Set default pull policy
+	if obj.PullPolicy == nil {
+		var pullPolicy corev1.PullPolicy
+		if obj.Tag != nil && *obj.Tag == "latest" {
+			pullPolicy = corev1.PullAlways
+		} else {
+			pullPolicy = corev1.PullIfNotPresent
+		}
+		obj.PullPolicy = &pullPolicy
+	}
+}
+
+func setDefaultsGardenlet(obj *Gardenlet, name, namespace string) {
+	// Set deployment defaults
+	if obj.Deployment == nil {
+		obj.Deployment = &GardenletDeployment{}
+	}
+
+	// Decode / initialize gardenlet config
+	var gardenletConfig *configv1alpha1.GardenletConfiguration
+	var err error
+	if obj.Config != nil {
+		// Decode gardenlet config to an external version
+		// Without defaults, since we don't want to set gardenlet config defaults in the resource at this point
```

Comment:
But this is only my opinion, if you wish we can go ahead and merge it from my PoV.

Reply:
I am not quite sure what exactly is the issue that you see, and what do you propose? 

The comment explains that we don't want all the defaults of the `GardenletConfiguration` type itself at this point. As was discussed, this is because it would break merging with parent, also as @timebertt pointed out it's not idiomatic, other Kubernetes implementations (e.g. `Deployment`) don't apply defaults on embedded types as well. But I think a full comprehensive explanation would result in a rather biggish comment that is not appropriate here.

As for Line 108, what happens there is that particular `GardenletConfiguration` defaults, that are specific to the managed seed, are applied. This is quite independent of the way `GardenletConfiguration` is decoded.